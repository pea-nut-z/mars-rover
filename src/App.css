html,
body {
  background: url("images/mars.jpeg") left bottom no-repeat;
  background-size: cover;
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  overflow: auto;
  color: white;
  font-size: 15px;
  font-family: "Montserrat", sans-serif;
}

:root {
  --clr-light: #fff;
  --clr-accent: #d06d6d;
  --pad-normal: 30px;
  --margin-normal: 2%;
  --white-background: linear-gradient(rgba(255, 255, 255, 0.1), rgba(255, 255, 255, 0.1));
  --black-background: linear-gradient(rgba(0, 0, 0, 0.3), rgba(0, 0, 0, 0.3));
}

p {
  margin: 0;
}

.weather-section {
  background: var(--white-background);
  width: 45%;
  padding: var(--pad-normal);
  margin: var(--margin-normal);
}

.image-section {
  background: var(--black-background);
  width: 45%;
  padding: var(--pad-normal);
  margin: var(--margin-normal);
}

.news-section {
  position: absolute;
  background: var(--black-background);
  width: 49%;
  height: 94%;
  padding: var(--pad-normal);
  margin: var(--margin-normal);
  top: 0;
  left: 47%;
  overflow: auto;
}

.likedImages-section {
  width: 100%;
  float: left;
}

.section-title {
  color: var(--clr-accent);
  font-weight: 500;
  font-size: 21px;
  padding-bottom: var(--pad-normal);
}

.divider {
  width: 5px;
  margin-left: var(--pad-normal);
  margin-right: var(--pad-normal);
  background-color: var(--clr-accent);
}

.openModalBtn {
  border: none;
  background-color: transparent;
}

.closeModalBtn {
  background-color: var(--clr-accent);
  margin: 10px;
  border-radius: 10px;
  margin-left: 45%;
}

.blackBtn {
  background-color: black;
  width: 150px;
  border: solid 2px white;
  color: white;
  margin: 1.5%;
  border-radius: 10px;
}

.heartBtn {
  border: none;
  background-color: transparent;
  color: var(--clr-accent);
  z-index: 1;
  position: relative;
  animation: fave-heart 1s steps(28);
}

.heart {
  stroke: white;
  stroke-width: 2;
}
button:focus {
  outline: none;
  text-decoration: none;
}

.unit {
  -ms-grid-column: 3;
  -ms-grid-column-span: 1;
  grid-column: 3 / 4;
  place-self: end;
  color: var(--clr-light);
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  opacity: 0.7;
  -webkit-transition: opacity 275ms linear;
  transition: opacity 275ms linear;
}

.unit:hover {
  opacity: 1;
}

.unit label {
  cursor: pointer;
}

.unit input {
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.unit-toggle {
  cursor: pointer;
  width: 4em;
  border: 2px solid var(--clr-light);
  background: transparent;
  padding: 0;
  border-radius: 100vmax;
  margin: 0 1em;
}

.unit-toggle::after {
  content: "";
  display: block;
  background: var(--clr-accent);
  border-radius: 50%;
  height: 1rem;
  margin: 1px;
  margin-left: auto;
  width: 1rem;
}

:checked ~ .unit-toggle::after {
  margin-left: 3px;
}

.unit-toggle:focus {
  outline: none;
  text-decoration: none;
}

.slide {
  width: 335px;
  height: 335px;
  border: solid 2px white;
  border-radius: 5px;
}

.carousel-control-prev-icon,
.carousel-control-next-icon {
  height: 100px;
  width: 100px;
  background-image: none;
}

.carousel-control-next-icon:after {
  content: ">";
  font-size: 55px;
  color: var(--clr-accent);
}

.carousel-control-prev-icon:after {
  content: "<";
  font-size: 55px;
  color: var(--clr-accent);
}
.carousel-indicators li {
  background-color: var(--clr-accent);
}

.news-images {
  margin-bottom: var(--pad-normal);
  margin-right: var(--pad-normal);
  border: solid 2px white;
  border-radius: 5px;
  width: 250px;
  height: 160px;
}

.news-images:hover {
  border: solid 2px var(--clr-accent);
}

::-webkit-scrollbar {
  width: 20px;
}

::-webkit-scrollbar-track {
  box-shadow: inset 0 0 5px grey;
  border-radius: 10px;
}

::-webkit-scrollbar-thumb {
  background: var(--clr-accent);
  border-radius: 10px;
}

.likedImageBackground {
  background: var(--black-background);
  padding: var(--pad-normal);
  width: 360px;
  height: 500px;
  display: flex;
  flex-direction: column;
  text-align: center;
  align-items: center;
  margin: 1%;
}

/* The container must be positioned relative: */
.custom-select {
  position: relative;
  font-family: Arial;
}

.custom-select select {
  display: none; /*hide original SELECT element: */
}

.select-selected {
  background-color: DodgerBlue;
}

/* Style the arrow inside the select element: */
.select-selected:after {
  position: absolute;
  content: "";
  top: 14px;
  right: 10px;
  width: 0;
  height: 0;
  border: 6px solid transparent;
  border-color: #fff transparent transparent transparent;
}

/* Point the arrow upwards when the select box is open (active): */
.select-selected.select-arrow-active:after {
  border-color: transparent transparent #fff transparent;
  top: 7px;
}

/* style the items (options), including the selected item: */
.select-items div,
.select-selected {
  color: #ffffff;
  padding: 8px 16px;
  border: 1px solid transparent;
  border-color: transparent transparent rgba(0, 0, 0, 0.1) transparent;
  cursor: pointer;
}

/* Style items (options): */
.select-items {
  position: absolute;
  background-color: DodgerBlue;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 99;
}

/* Hide the items when the select box is closed: */
.select-hide {
  display: none;
}

.select-items div:hover,
.same-as-selected {
  background-color: rgba(0, 0, 0, 0.1);
}

@media (max-width: 576px) {
  body {
    font-size: 14px;
    overflow: auto;
  }

  .unit {
    position: relative;
    left: 6%;
  }

  .unit-toggle {
    width: 3em;
    border: 1px solid var(--clr-light);
    margin: 4px;
  }

  .unit-toggle::after {
    background: white;
  }

  .label {
    font-size: 13px;
    margin: 0;
    padding: 0;
  }

  .weather-section {
    width: 100%;
    margin: 0;
  }

  .weather-container {
    flex-direction: column;
  }

  .section-title {
    width: 50%;
  }

  /* IMAGES */
  .image-section {
    width: 100%;
    margin: 0;
  }

  .image-title {
    text-align: center;
  }

  .slide {
    width: 270px;
    height: 270px;
    border-radius: 5px;
  }

  /* NEWS */
  .news-section {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center !important;
    left: 0;
    margin: 0;
    width: 100%;
    background: var(--white-background);
  }

  .news-container {
    display: flex;
    flex-direction: column !important;
  }
  .news-images {
    width: 240px;
    height: 150px;
    margin-bottom: var(--margin-normal);
    margin-top: var(--pad-normal);
  }

  /* LIKES */
  .likedImageBackground {
    height: 440px;
  }
}

@media (max-width: 1450px) {
  body {
    overflow: auto;
  }
  .weather-section {
    width: 100%;
    margin: 0;
  }
  .image-section {
    position: relative;
    width: 100%;
    margin: 0;
  }
  .news-section {
    background: var(--white-background);
    position: relative;
    width: 100%;
    margin: 0;
    left: 0;
  }
}
