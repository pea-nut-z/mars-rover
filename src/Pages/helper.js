// FOR TESTING
let testYear, testMonth, testDay;

// CAMERAS
export const allCamerasAbb = [
  "FHAZ",
  "RHAZ",
  "MAST",
  "CHEMCAM",
  "MAHLI",
  "MARDI",
  "NAVCAM",
  "PANCAM",
  "MINITES",
];

export const allCameras = {
  1: "Front Hazard Avoidance Camera",
  2: "Rear Hazard Avoidance Camera",
  3: "Mast Camera",
  4: "Chemistry and Camera Complex",
  5: "Mars Hand Lens Imager",
  6: "Mars Descent Imager",
  7: "Navigation Camera",
  8: "Panoramic Camera",
  9: "Miniature Thermal Emission Spectrometer (Mini-TES)",
};

export const camerasAvailable = {
  curiosity: [1, 2, 3, 4, 5, 6, 7],
  opportunity: [1, 2, 7, 8, 9],
  spirit: [1, 2, 7, 8, 9],
};

export const curiosityCamerasAbb = ["RHAZ", "MAST", "CHEMCAM", "MAHLI", "MARDI", "NAVCAM"];

// FUNCTIONS
export const getTodayDate = () => {
  const date = new Date();
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  return [year, month, day];
};

export const getPreviousMonthDate = (year, month) => {
  const monthsWith30Days = [4, 6, 9, 11];
  const preMonth = month === 1 ? 12 : month - 1;
  let preYear = year;
  let preDay;

  if (preMonth === 2) {
    preDay = 28;
  } else if (monthsWith30Days.includes(preMonth)) {
    preDay = 30;
  } else {
    preYear = preMonth === 12 && year - 1;
    preDay = 31;
  }
  return [preYear, preMonth, preDay];
};

export const convertCelToFah = (cel) => {
  const fah = (cel * 9) / 5 + 32;
  return fah.toFixed(2);
};

export const validateOtherImgUrl = (url) => {
  const validUrls = [];
  const validRHAZUrl = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=RHAZ&earth_date=${testYear}-${testMonth}-${testDay}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;

  curiosityCamerasAbb.forEach((abbr) => {
    if (abbr === "RHAZ") return;
    const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=${abbr}&earth_date=${testYear}-${testMonth}-${testDay}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
    validUrls.push(url);
  });
  const result = validUrls.includes(url)
    ? "valid"
    : validRHAZUrl === url
    ? "valid & RHAZ"
    : undefined;
  return result;
};

export const filterLikes = (likedImages, imageId, roverName, imgUrl, cameraName, earthDate) => {
  let filtered;
  if (likedImages.hasOwnProperty(imageId)) {
    filtered = Object.keys(likedImages)
      .filter((id) => id != imageId)
      .reduce((obj, key) => {
        obj[key] = likedImages[key];
        return obj;
      }, {});
  } else {
    filtered = {
      ...likedImages,
      [imageId]: {
        roverName,
        imgUrl,
        cameraName,
        earthDate,
      },
    };
  }
  console.log({ filtered });

  return filtered;
};

// API
export const weatherUrl = "https://api.maas2.apollorion.com/";
export const getImgUrl = (cameraAbbr, dateArray) => {
  const [year, month, day] = dateArray;
  const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=${cameraAbbr}&earth_date=${year}-${month}-${day}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
  return url;
};
export const getFtImgUrl = (dateArray) => {
  [testYear, testMonth, testDay] = dateArray;
  const [year, month, day] = dateArray;
  const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=FHAZ&earth_date=${year}-${month}-${day}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
  return url;
};
// export const getOtherImgUrl = (cameraAbbr, dateArray) => {
//   const [year, month, day] = dateArray;
//   const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=${cameraAbbr}&earth_date=${year}-${month}-${day}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
//   return url;
// };

export const getFilteredImgUrl = (rover, camera, date) => {
  const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/${rover}/photos?camera=${camera}&earth_date=${date}&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
  return url;
};

export const newsUrl = `https://newsapi.org/v2/everything?qInTitle=%22mars%22&domains=nature.com&sortBy=publishedAt&apiKey=${process.env.REACT_APP_NEWS_API_KEY}`;

//STYLE
export const styleHeart = {
  stroke: "black",
  strokeWidth: "10",
};
