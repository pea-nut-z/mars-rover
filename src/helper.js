export const getTodayDate = () => {
  const date = new Date();
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  return [year, month, day];
};

export const camerasAbbr = ["RHAZ", "MAST", "CHEMCAM", "MAHLI", "MARDI", "NAVCAM"];
// export const camerasFullNames = [
//   "Rear Hazard Avoidance Camera",
//   "Mast Camera",
//   "Chemistry and Camera Complex",
//   "Mars Hand Lens Imager",
//   "Mars Descent Imager",
//   "Navigation Camera",
// ];

// API
export const weatherUrl = "https://api.maas2.apollorion.com/";
export const getFtImgUrl = (dateArray) => {
  const [year, month, day] = dateArray;
  const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=FHAZ&earth_date=${year}-${month}-${day}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
  return url;
};
export const getOtherImgUrl = (cameraAbbr, dateArray) => {
  const [year, month, day] = dateArray;
  const url = `https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?camera=${cameraAbbr}&earth_date=${year}-${month}-${day}&page=1&api_key=${process.env.REACT_APP_NASA_API_KEY}`;
  return url;
};
