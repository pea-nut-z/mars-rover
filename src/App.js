import React, { useState, useEffect, useRef } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { NavBar } from "./components/NavBar";
import Home from "./Pages/Home";
import Likes from "./Pages/Likes";
import Search from "./Pages/Search";
import "./App.css";
import * as func from "./helper";

export default function App() {
  const ref = useRef();

  const closeNavBar = () => {
    ref.current.navBar && ref.current.toggleNavBar();
  };

  return (
    <div onClick={closeNavBar}>
      <Router>
        <NavBar ref={ref} />
        <Switch>
          <Route exact path="/" component={Home} />
          <Route path="/likes" component={Likes} />
          <Route path="/search" component={Search} />
        </Switch>
      </Router>
    </div>
  );
}
