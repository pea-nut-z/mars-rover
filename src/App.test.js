import React from "react";
import "@testing-library/jest-dom";
import renderer, { act } from "react-test-renderer";
import axios from "axios";
import App from "./App";
import { fakeWeatherData, fakeFtImgData } from "./fake.data";
import { weatherUrl, getFtImgUrl, getOtherImgUrl, getTodayDate, camerasAbbr } from "./helper";

import Modal from "react-modal";

jest.mock("axios");

const dateArray = getTodayDate();
const frontCameraImgUrl = getFtImgUrl(dateArray);
const fakeRHAZImgUrl = getOtherImgUrl(cameraAbbr, dateArray);

const axiosGetSpy = jest.spyOn(axios, "get").mockImplementation((url) => {
  switch (url) {
    case weatherUrl:
      return Promise.resolve({ data: fakeWeatherData });
    case frontCameraImgUrl:
      return Promise.resolve({ data: { photos: fakeFtImgData } });
    default:
      return Promise.reject(new Error("not found"));
  }
});

describe("Fetch Data", () => {
  let component;
  beforeAll(async () => {
    await act(() => {
      return axios.get(weatherUrl).then(() => {
        component = renderer.create(<App />);
      });
    });
  });

  afterAll(() => {
    component.unmount();
    axiosGetSpy.mockRestore();
  });

  it("fetches from correct url for weather data.", async () => {
    expect(axiosGetSpy).toBeCalledWith(weatherUrl);
  });

  it("fetches from correct url with queries for image data.", async () => {
    // expect(axiosGetSpy).toBeCalledWith(frontCameraImgUrl);

    await act(() => {
      return axios.get(frontCameraImgUrl).then((res) => {
        console.log(res);
      });
    });
  });
});

// describe("Layout", () => {

//   afterAll(() => {
//     component.unmount();
//     axiosGetSpy.mockRestore();
//   });

//   it("fetches and renders weather data.", async () => {
//     let component;
//     await act(() => {
//       return axios.get("https://api.maas2.apollorion.com/").then(() => {
//         component = renderer.create(<App />);
//       });
//     });
//     expect(component.toJSON()).toMatchSnapshot();
//   });
// });
