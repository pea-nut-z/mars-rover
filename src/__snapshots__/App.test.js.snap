// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout fetches and renders weather data. 1`] = `
<div>
  <div
    className={undefined}
  />
  <div
    className="weather-section"
  >
    <div
      className="d-flex justify-content-between"
    >
      <div
        className="section-title"
      >
        LATEST WEATHER AT ELYSIUM PLANTITIA
      </div>
      <div
        className="unit"
      >
        <label
          className="label h5"
          htmlFor="cel"
        >
          째C
        </label>
        <input
          checked="checked"
          id="cel"
          name="unit"
          readOnly={true}
          type="radio"
        />
        <button
          className="unit-toggle"
          onClick={[Function]}
        />
        <label
          className="label h5"
          htmlFor="fah"
        >
          째F
        </label>
        <input
          checked={false}
          id="fah"
          name="unit"
          readOnly={true}
          type="radio"
        />
      </div>
    </div>
    <div
      className="weather-container d-flex"
    >
      <div
        className="text-right"
      >
        <div
          className="sol display-4"
        >
          Sol 
          1314
        </div>
        <div
          className="h2 text-muted"
        >
          Jun
           
          30
        </div>
      </div>
      <div
        className="divider"
      />
      <div
        className="col"
      >
        <div
          className="h5"
        >
          High: 
          -25째C
        </div>
        <div
          className="h5"
        >
          Low: 
          -78째C
        </div>
        <div>
          <div>
            Wind Speed: 
            N/A
          </div>
          <div>
            Wind Direction: 
            N/A
          </div>
        </div>
      </div>
      <div
        className="col text-right"
      >
        <div
          className="h5"
        >
          Sunny
        </div>
        <div>
          <div>
            Autumn
          </div>
          <div>
            UV Index: 
            Moderate
          </div>
          <div>
            Sunrise: 
            06:02
          </div>
          <div>
            Sunset: 
            17:47
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
